main.go

```diff
@@ -106,11 +120,191 @@ func newVault(addr string) (vault *vapi.Client) {
+
+func getVaultAddr(service []core.ServicePort, ip string) string {
+	const targetScheme = "http"
+
+	for _, port := range service {
+		if port.Name == targetScheme {
+			return fmt.Sprintf("%s://%s:%d", port.Protocol, ip, port.Port)
+		}
+	}
+
+	return fmt.Sprintf("%s://%s:8200", targetScheme, ip) // Default to 8200 on http.
+}
+
+// monitorService is a function that will monitor a service for changes to the pods attached to it.
+func monitorService(ctx context.Context, workerIps *sync.Map, wg *sync.WaitGroup) {
+	client, err := getKubeClient()
+	if err != nil {
+		logger.WithError(err).Fatal("error getting kubernetes client")
+	}
+
+	ticker := time.NewTicker(defaultMonitoringInterval)
+
+	for range ticker.C {
+		select {
+		case <-ctx.Done():
+			logger.Info("closing service monitor")
+			return
+		default:
+			services, err := client.CoreV1().Services(getNamespaceFromConfig()).List(context.Background(), metav1.ListOptions{
+				LabelSelector: labels.Set{
+					appNameLabel: defaultVaultName,
+				}.String(),
+			})
+			if err != nil {
+				logger.WithError(err).Error("error getting services")
+				continue
+			}
+
+			if len(services.Items) == 0 {
+				logger.Warn("no services found")
+				continue
+			}
+
+			fmt.Println(workerIps)
+
+			svcName := strings.Split(conf.VaultService, ".")[1]
+			for _, service := range services.Items {
+				if service.Name != svcName {
+					continue
+				}
+
+				pods, err := client.CoreV1().Pods(getNamespaceFromConfig()).List(context.Background(), metav1.ListOptions{
+					LabelSelector: labels.Set{
+						appNameLabel:  defaultVaultName,
+						instanceLabel: service.Name,
+					}.String(),
+				})
+				if err != nil {
+					logger.WithError(err).Errorf("error getting pods for service %q", service.Name)
+					continue
+				}
+
+				// Check for new pods.
+				for _, pod := range pods.Items {
+					addr := getVaultAddr(service.Spec.Ports, pod.Status.PodIP)
+					if _, ok := workerIps.Load(addr); !ok {
+						logger.WithField("addr", addr).Info("adding worker")
+						wg.Add(1)
+						workerCtx, workerCancel := context.WithCancel(ctx)
+						workerIps.Store(addr, workerCancel)
+						go worker(workerCtx, wg, addr)
+					}
+				}
+
+				// Check for removed pods.
+				workerIps.Range(func(key, value interface{}) bool {
+					addr, ok := key.(string)
+					if !ok {
+						// This should never happen. Do not stop the loop.
+						return true
+					}
+
+					found := false
+					for _, pod := range pods.Items {
+						podAddr := getVaultAddr(service.Spec.Ports, pod.Status.PodIP)
+						if addr == podAddr {
+							found = true
+							break
+						}
+					}
+
+					if !found {
+						logger.WithField("addr", addr).Info("removing worker")
+						value.(context.CancelFunc)()
+						workerIps.Delete(addr)
+					}
+
+					return true
+				})
+			}
+
+			fmt.Println(workerIps)
+		}
+	}
+}
+
+// getNamespaceFromConfig returns the namespace from the config.
+//
+// E.g. name.service.namespace.svc.cluster.local:some-port
```

Comment:
isn't the address "svc-name.namespace.svc.cluster.local"?

Reply:
No because we are wanting to access the pods IP directly.